product/cmd/grpc/client/main_test.go

```diff
@@ -0,0 +1,58 @@
+package main
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/stretchr/testify/require"
+
+	"product/internal/grpc/domain"
+	domainmock "product/internal/grpc/domain/mock_domain"
+)
+
+var product = &domain.Product{
+	Id:          1,
+	Name:        "Sherlock Holmes",
+	Description: "It is an investigative book",
+	Price:       123.56,
+	Slug:        "sherlock-holmes",
+	Quantity:    60,
+	CreatedAt:   &domain.Timestamp{Timestamp: ptypes.TimestampNow()},
+}
+
+var req = &domain.AddProductRequest{
+	Product: product,
+}
+
+var resp = &domain.AddProductResponse{
+	CreatedProduct: product,
+}
+
+func TestAddProduct(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	mockProdServiceClient := domainmock.NewMockProductServiceClient(ctrl)
```

Comment:
I very strongly recommend against using mocks for testing. The simplest rationale I can give is that it's a circular argument that only proves that your code calls the functions that you tell your code to call. For example:

```
type dep interface {
  bar() int
}

// foo returns dep.bar() + 42. (In your case this would be AddProduct adds a product)
func foo(d dep) int {
  return dep.bar() + 42
}
```

```
func TestFoo() {
  …
  mock = depmock.NewMocked(ctrl)
  mock.EXPECT().bar()
  …
}
```

What does this test prove? That you called `bar()`! Why do you want to prove that you called bar? Because that's what you implemented in the code (circular argument). But does it prove that it works? No, it proves that it does what it does, and if you change what it does then you have to change the test if you achieve the same thing in a different way. This is called a "change detector test" because all it detects is that you changed your implementation. What you really want to detect is that you broke your code! The problem is not only that you always have to change your tests, but that you never know if you're updating the tests because of the change or because you actually broke something.

You should prefer to write behaviour-driven tests. My function says that it has some well-described behaviour, so I'm going to check exactly that. Call your gRPC service's functions and check that it returns the correct value (like you do with `cmp.Diff`, just like if you were writing a regular function. If there are side effects (e.g. new product in the database) then check that those have happened but, even better, check them via the `ListProducts()` RPC function (call it and see that there are no products, then call `AddProduct()`, then call `ListProducts()` again).